summingbird-storm/src/main/scala/com/twitter/summingbird/storm/collector/AggregatorOutputCollector.scala

```diff
@@ -0,0 +1,122 @@
+
+  // The Map keeps track of batch of aggregated tuples' messageIds. It also has a stream level tracking.
+  private val streamMessageIdTracker = MMap[String, MMap[Int, MList[Object]]]()
+
+  def timerFlush() = {
+    /*
+    This is a flush called from the nextTuple() of the spout.
+    The timerFlush is triggered with tick frequency from the spout.
+     */
+    spoutCaches.foreach {
+      case (stream, cache) =>
+        val tupsOut = cache.tick.map { convertToSummerInputFormat(_) }
+        emitData(tupsOut, stream)
+    }
+  }
+
+  private def convertToSummerInputFormat(flushedCache: CMap[K, V]): CMap[Int, CMap[K, V]] =
+    flushedCache.groupBy { case (k, _) => summerShards.summerIdFor(k) }
+
+  /*
+    The method is invoked to handle the flushed cache caused by
+    exceeding the memoryLimit, which is called within add method.
+   */
+  private def emitData[K, V](cache: Future[TraversableOnce[(Int, CMap[K, V])]], streamId: String): JList[Integer] = {
+    val flushedTups = Await.result(cache)
+    val messageIdsTracker = streamMessageIdTracker(streamId)
+    val returns = flushedTups.map {
+      case (k, v) =>
+        val messageIds = messageIdsTracker.remove(k)
+        val list = List(k, v).asJava.asInstanceOf[JList[AnyRef]]
+        callEmit(messageIds, list, streamId)
+    }
+    val result = new java.util.ArrayList[Integer]()
+    returns.filter(_ != null).foreach { result.addAll(_) }
+    result
+  }
+
+  /*
+   This is a wrapper method to call the emit with appropriate signature
+   based on the arguments.
+  */
+  private def callEmit(messageIds: Option[TraversableOnce[AnyRef]], list: JList[AnyRef], stream: String): JList[Integer] = {
+    (messageIds.isEmpty, stream.isEmpty) match {
+      case (true, true) => in.emit(list)
+      case (true, false) => in.emit(stream, list)
+      case (false, true) => in.emit(list, messageIds)
+      case (false, false) => in.emit(stream, list, messageIds)
+    }
+  }
+
+  /*
+  Method wraps the adding the tuple to the spoutCache along with adding the corresponding
+  messageId to the messageId Tracker.
+   */
+  private def add(tuple: (K, V), streamid: String, messageId: Option[AnyRef] = None) = {
+    if (messageId.isDefined)
+      trackMessageId(tuple, messageId.get, streamid)
+    addToCache(tuple, streamid)
+  }
+
+  private def addToCache(tuple: (K, V), streamid: String) = {
+    spoutCaches.get(streamid) match {
+      case Some(cac) => cac.add(tuple)
+      case None => {
+        spoutCaches(streamid) = summerBuilder.getSummer[K, V](implicitly[Semigroup[V]])
+        spoutCaches(streamid).add(tuple)
+      }
+    }
+  }
+
+  private def trackMessageId(tuple: (K, V), o: AnyRef, s: String): Unit = {
+    val messageIdTracker = streamMessageIdTracker.getOrElse(s, MMap[Int, MList[Object]]())
+    var messageIds = messageIdTracker.getOrElse(summerShards.summerIdFor(tuple._1), MList())
+    messageIdTracker(summerShards.summerIdFor(tuple._1)) = messageIds += o
```

Comment:
We aggregate the tuples based on key's summer shard. So we aggregate the messageIds on the summer shards, so that we can emit the aggregated tuple along with the aggregate messageIds.